tests/ga/test_policy_engine.py

```diff
@@ -15,20 +15,406 @@
 # Requires Python 2.4+ and Openssl 1.0+
 #
 
+import os
+import json
+
 from tests.lib.tools import AgentTestCase
-from azurelinuxagent.ga.policy.policy_engine import PolicyEngine
+from azurelinuxagent.ga.policy.policy_engine import ExtensionPolicyEngine, PolicyInvalidError
 from tests.lib.tools import patch
+from azurelinuxagent.common.protocol.restapi import Extension
+
+TEST_EXTENSION_NAME = "Microsoft.Azure.ActiveDirectory.AADSSHLoginForLinux"
+
+
+class _TestPolicyBase(AgentTestCase):
+    """
+    Define common methods for policy engine test classes.
+    """
+    def setUp(self):
+        AgentTestCase.setUp(self)
+        self.custom_policy_path = os.path.join(self.tmp_dir, "waagent_policy.json")
+
+        # Patch attributes to enable policy feature
+        self.patch_custom_policy_path = patch('azurelinuxagent.ga.policy.policy_engine._CUSTOM_POLICY_PATH',
+                                              new=self.custom_policy_path)
+        self.patch_custom_policy_path.start()
+        self.patch_conf_flag = patch('azurelinuxagent.ga.policy.policy_engine.conf.get_extension_policy_enabled',
+                                     return_value=True)
+        self.patch_conf_flag.start()
+
+    def tearDown(self):
+        patch.stopall()
+        AgentTestCase.tearDown(self)
+
+    def _create_policy_file(self, policy):
+        with open(self.custom_policy_path, mode='w') as policy_file:
+            json.dump(policy, policy_file, indent=4)
+            policy_file.flush()
+
 
+class TestPolicyEngine(_TestPolicyBase):
```

Comment:
can we add unit tests for when user provides invalid types for the following attributes in the policy:
- "extensions"
- key/value pairs in "extensions"

Reply:
Added tests for these and a few more edge cases